aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -166,6 +166,44 @@ SparseTensor& asin_sparse_(SparseTensor& t) {
   return asin_out_sparse(t, t);
 }
 
+// --------------------------------------------------------------------
+// asinh(SparseTensor)
+// --------------------------------------------------------------------
+
+// In-place asinh on uncoalesced tensors is not supported since the operation is
+// not a linear map. Values of uncoalesced tensor corresponding to the same
+// indices are summed and asinh(summed_value) != asinh(v1) + asinh(v2)
+
+SparseTensor& asinh_out_sparse(SparseTensor& r, const SparseTensor& t) {
+  TORCH_CHECK(r.is_sparse(), "Tensor should be sparse");
+  TORCH_CHECK(t.is_sparse(), "Tensor should be sparse");
+  TORCH_CHECK(
```

Comment:
I suggest writing a helper function to validate the inputs to sparse unary out operations.

The helper needs to check:

- that both tensors are sparse
- that both tensors are on the same device
- that the output of the unary operation can be safely cast to the out tensor's dtype
  - this requires understanding the output dtype, which is typically the input's dtype but, in cases like asinh, may be the default scalar type if the input is an integer (incl. bool) type
  - a flag on the helper for whether the function promotes integer (incl. bool) inputs to float or not would help with this 
- that if the input and out tensor are the same tensor then the input is coalesced 
  -  ... unless the operation is a linear map, which can be another flag passed to the helper 
- that either the out tensor either has no elements, in which case it should be resized to hold the output at some point, or that it is the correct size to hold the output
  - note this requires understanding the output size, which probably means coalescing the input

This helper function can then be reused by other sparse unary ops. Check out UnaryOps.cpp, in particular functions like

https://github.com/pytorch/pytorch/blob/4a3a37886cdbdca108536a2730c034db4f96f095/aten/src/ATen/native/UnaryOps.cpp#L69

to get a sense for what the helper's look like. This current dtype check wouldn't deal with complex inputs properly, for example, because they cannot be safely cast to float. Typically safe casting and resizing logic is implemented in TensorIterator, but there's not currently a TensorIterator for sparse operations, so some of its logic needs to be duplicated here.

Reply:
@mruberry - I have put in a helper function for doing the following checks:

- that both tensors are sparse

- that both tensors are on the same device
- that if the input and out tensor is the same tensor then the input has coalesced
... unless the operation is a linear map, which can be another flag passed to the helper

I have not added the check for type as there is a utility function get_result_tensor_for_unary_op which will generate the required result tensor; have put the the promotes_integer_to_float flag for checking the dtype for that particular case.